iroha/src/wsv.rs

```diff
@@ -42,187 +41,199 @@ impl WorldStateView {
         for block in blocks {
             self.apply(&block.clone().commit());
         }
     }
 
     /// Apply `CommittedBlock` with changes in form of **Iroha Special Instructions** to `self`.
-    pub fn apply(&mut self, block: &VersionedCommittedBlock) {
+    pub fn apply(&self, block: &VersionedCommittedBlock) {
         for transaction in &block.as_inner_v1().transactions {
             if let Err(e) = transaction.proceed(self) {
                 iroha_logger::warn!("Failed to proceed transaction on WSV: {}", e);
             }
-            let _ = self.transactions.insert(
+            drop(self.transactions.insert(
                 transaction.hash(),
                 TransactionValue::Transaction(transaction.clone().into()),
-            );
+            ));
         }
         for transaction in &block.as_inner_v1().rejected_transactions {
-            let _ = self.transactions.insert(
+            drop(self.transactions.insert(
                 transaction.hash(),
                 TransactionValue::RejectedTransaction(transaction.clone()),
-            );
+            ));
         }
     }
 
     /// Get `World` without an ability to modify it.
-    pub const fn read_world(&self) -> &World {
+    pub const fn world(&self) -> &World {
         &self.world
     }
 
-    /// Get `World` with an ability to modify it.
-    pub fn world(&mut self) -> &mut World {
-        &mut self.world
-    }
-
     /// Add new `Domain` entity.
     pub fn add_domain(&mut self, domain: Domain) {
-        let _ = self.world.domains.insert(domain.name.clone(), domain);
+        drop(self.world.domains.insert(domain.name.clone(), domain));
     }
 
-    /// Get `Domain` without an ability to modify it.
-    pub fn read_domain(&self, name: &str) -> Option<&Domain> {
-        self.world.domains.get(name)
+    /// Returns reference for domains map
+    pub const fn domains(&self) -> &DomainsMap {
+        &self.world.domains
     }
 
-    /// Get `Domain` with an ability to modify it.
-    pub fn domain(&mut self, name: &str) -> Option<&mut Domain> {
-        self.world.domains.get_mut(name)
+    /// Returns reference for trusted peer ids
+    pub const fn trusted_peers_ids(&self) -> &PeersIds {
+        &self.world.trusted_peers_ids
     }
 
-    /// Get all `Domain`s without an ability to modify them.
-    pub fn read_all_domains(&self) -> Vec<&Domain> {
-        let mut vec = self.world.domains.values().collect::<Vec<&Domain>>();
-        vec.sort();
-        vec
+    /// Returns reference for parameters
+    pub const fn parameters(&self) -> &RwLock<Vec<Parameter>> {
+        &self.world.parameters
     }
 
-    /// Get all `Account`s without an ability to modify them.
-    pub fn read_all_accounts(&self) -> Vec<&Account> {
-        let mut vec = self
+    /// Get `Domain` without an ability to modify it.
+    /// # Errors
+    /// Fails if there is no domain
+    pub fn domain<T>(&self, name: &str, f: impl FnOnce(&Domain) -> T) -> Result<T> {
```

Comment:
I think you mentioned this, but just to be clear, you suggest to use lambdas not to return Guards?

Reply:
There is a bunch of neat problems related to rust type system. I guess we could do our own wrapper, but it would require working with rust contracts, lifetimes and unsafe code. Using lambdas is just faster for this iteration.